src/function/table/arrow_conversion.cpp

```diff
@@ -120,55 +120,124 @@ static void ColumnArrowToDuckDBDictionary(Vector &vector, ArrowArray &array, Arr
                                           idx_t size, const ArrowType &arrow_type, int64_t nested_offset = -1,
                                           ValidityMask *parent_mask = nullptr, uint64_t parent_offset = 0);
 
-static void ArrowToDuckDBList(Vector &vector, ArrowArray &array, ArrowArrayScanState &array_state, idx_t size,
-                              const ArrowType &arrow_type, int64_t nested_offset, ValidityMask *parent_mask,
-                              int64_t parent_offset) {
-	auto size_type = arrow_type.GetSizeType();
-	idx_t list_size = 0;
-	auto &scan_state = array_state.state;
+namespace {
 
-	SetValidityMask(vector, array, scan_state, size, parent_offset, nested_offset);
+struct ArrowListOffsetData {
+	idx_t list_size = 0;
 	idx_t start_offset = 0;
+};
+
+} // namespace
+
+template <class BUFFER_TYPE>
+static ArrowListOffsetData ConvertArrowListOffsetsTemplated(Vector &vector, ArrowArray &array, idx_t size,
+                                                            idx_t effective_offset) {
+	ArrowListOffsetData result;
+	auto &start_offset = result.start_offset;
+	auto &list_size = result.list_size;
+
 	idx_t cur_offset = 0;
+	auto offsets = ArrowBufferData<BUFFER_TYPE>(array, 1) + effective_offset;
+	start_offset = offsets[0];
+	auto list_data = FlatVector::GetData<list_entry_t>(vector);
+	for (idx_t i = 0; i < size; i++) {
+		auto &le = list_data[i];
+		le.offset = cur_offset;
+		le.length = offsets[i + 1] - offsets[i];
+		cur_offset += le.length;
+	}
+	list_size = offsets[size];
+	list_size -= start_offset;
+	return result;
+}
+
+template <class BUFFER_TYPE>
+static ArrowListOffsetData ConvertArrowListViewOffsetsTemplated(Vector &vector, ArrowArray &array, idx_t size,
+                                                                idx_t effective_offset) {
+	ArrowListOffsetData result;
+	auto &start_offset = result.start_offset;
+	auto &list_size = result.list_size;
+
+	list_size = 0;
+	auto offsets = ArrowBufferData<BUFFER_TYPE>(array, 1) + effective_offset;
+	auto sizes = ArrowBufferData<BUFFER_TYPE>(array, 2) + effective_offset;
+
+	// In ListArrays the offsets have to be sequential
+	// ListViewArrays do not have this same constraint
+	// for that reason we need to keep track of the lowest offset, so we can skip all the data that comes before it
+	// when we scan the child data
+	auto lowest_offset = offsets[0];
+	auto list_data = FlatVector::GetData<list_entry_t>(vector);
+	for (idx_t i = 0; i < size; i++) {
+		auto &le = list_data[i];
+		le.offset = offsets[i];
+		le.length = sizes[i];
+		list_size += le.length;
+		lowest_offset = MinValue(lowest_offset, offsets[i]);
+	}
```

Comment:
I recommend creating a dedicated function just to calculating the lowest offset (see version I wrote in Arrow [1]). A good optimization is to early-return if `0` is found because you can't possibly go lower than `0` on offsets and `0` will be very commonly the first offset in practice for most vectors.

[1] https://github.com/apache/arrow/blob/main/cpp/src/arrow/util/list_util.cc#L45

Reply:
Thanks I'll have a look at that 👍

Reply:
I have decided to not make a separate function
As you mentioned nulls are bound to the same invariants, so we can expect them to never be out of bounds
Worst case they'll be too low and we scan too much of the child array, which shouldn't break anything.

I have added the condition that the size needs to be over 0 to be registered as a valid offset, thanks for the suggestion.